graphql-dgs-client/src/main/kotlin/com/netflix/graphql/dgs/client/WebsocketGraphQLClient.kt

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2021 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs.client
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.annotation.JsonSubTypes
+import com.fasterxml.jackson.annotation.JsonTypeInfo
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import graphql.GraphQLException
+import org.reactivestreams.Publisher
+import org.springframework.web.reactive.socket.WebSocketMessage
+import org.springframework.web.reactive.socket.WebSocketSession
+import org.springframework.web.reactive.socket.client.WebSocketClient
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.core.publisher.Sinks
+import java.net.URI
+
+/**
+ * Reactive client implementation using websockets and the subscription-transport-ws protocol:
+ * https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md
+ */
+class WebsocketGraphQLClient(
+    private val url: String,
+    private val client: WebSocketClient): ReactiveGraphQLClient {
+    companion object {
+        private val CONNECTION_INIT_MESSAGE = OperationMessage(GQL_CONNECTION_INIT, null, null)
+        private val CONNECTION_TERMINATE_MESSAGE = OperationMessage(GQL_CONNECTION_TERMINATE, null, null)
+        private val MAPPER = jacksonObjectMapper()
+    }
+
+    override fun reactiveExecuteQuery(
+        query: String,
+        variables: Map<String, Any>,
+    ): Publisher<GraphQLResponse> {
+        return reactiveExecuteQuery(query, variables, null)
+    }
+
+    override fun reactiveExecuteQuery(
+        query: String,
+        variables: Map<String, Any>,
+        operationName: String?,
+    ): Publisher<GraphQLResponse> {
+        val uri = URI(url)
+
+        // Sinks are used as buffers, incoming messages from the server are
```

Comment:
I believe an implementation without sinks should be possible. If you want to buffer due to back pressure, I'd recommend using one of the operators on the stream.

Reply:
Because `client.execute` must return a `Mono<Void>`, the alternative to sinks is something like

```kotlin
return Flux.create {
    emitter -> client.execute(uri) { session -> exchange(emitter, session) }
}
```

Would that be preferred?

Reply:
ah, yes, you're correct. i'm not sure that is necessarily better.

what i _would_ try to do is remove the naked `client.subscribe()` call and tie that into the overall flow. (perhaps by wrapping everything inside of the method in a `Flux.defer`)

the returned `Publisher` of this method should be able to be subscribed to multiple times without re-invoking `reactiveExecuteQuery`. the connection shouldn't be active until the returned `Publisher` is subscribed to.

Reply:
Have created a single `Mono` for connection to the server which is cached to fulfil https://github.com/Netflix/dgs-framework/pull/522#issuecomment-895369056, the mono is chained to the flux of messages about a particular subscription, it is also deferred so connection does not occur until the first time the user calls `subscribe`